### PR TITLE
Set NotNull constraints for object creations in SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5860

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 1]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+{
+    var value = new T();
+    Tag(""value"", value);
+}
+
+private void Tag(string name, object arg) { }";
+            var validator = SETestContext.CreateCSMethod(code).Validator;
+            validator.ValidateContainsOperation(OperationKind.TypeParameterObjectCreation);
+            validator.ValidateTag("value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:560]: This is probably prematurely but what about `var x = condition ? new object() : null;`?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 2]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+        public void TypeParameterObjectCreation_SetsNotNull()
+        {
+            const string code = @"
+public void Main<T>() where T : new()
+{
+    var value = new T();
+    Tag(""value"", value);
+}
+
+private void Tag(string name, object arg) { }";
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:556]: Maybe better:
```suggestion
private void Tag(string name, T arg) { }";
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 3]
```diff
@@ -244,14 +244,20 @@ semantics switch
...
+                OperationKindEx.DynamicObjectCreation => Creation.Process(context),
                 OperationKindEx.FieldReference => References.Process(context, IFieldReferenceOperationWrapper.FromOperation(context.Operation.Instance)),
                 OperationKindEx.FlowCapture => FlowCapture.Process(context, IFlowCaptureOperationWrapper.FromOperation(context.Operation.Instance)),
                 OperationKindEx.IsPattern => Pattern.Process(context, IIsPatternOperationWrapper.FromOperation(context.Operation.Instance)),
                 OperationKindEx.LocalReference => References.Process(context, ILocalReferenceOperationWrapper.FromOperation(context.Operation.Instance)),
+                OperationKindEx.ObjectCreation => Creation.Process(context),
                 OperationKindEx.ParameterReference => References.Process(context, IParameterReferenceOperationWrapper.FromOperation(context.Operation.Instance)),
                 OperationKindEx.SimpleAssignment => SimpleAssignment.Process(context, ISimpleAssignmentOperationWrapper.FromOperation(context.Operation.Instance)),
+                OperationKindEx.TypeParameterObjectCreation => Creation.Process(context),
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:260]: Just to make sure: Are these all ..ObjectCreation operations?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 4]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+            validator.ValidateTag("arr3", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("arrMulti", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("arrJagged", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
+
+        [TestMethod]
+        public void DelegateCreation_SetsNotNull()
+        {
+            const string code = @"
+var pointer = Main; // Delegate creation to encapsulating method
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:509]: What about
```cs
var d = () => M();
var d = delegate() {};
```
Should this also be considered a "creation"?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 5]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+        public void ArrayCreation_SetsNotNull()
+        {
+            const string code = @"
+var arr1 = new int[] { 42 };
+var arr2 = new int[0];
+int[] arr3 = { };
+int[,] arrMulti = new int[2, 3];
+int[][] arrJagged = new int[2][];
+
+Tag(""arr1"", arr1);
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:491]: Suggestion, We could use `[CallerMemberName]` to get rid of the `""arr1""`.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 6]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+            validator.ValidateContainsOperation(OperationKind.DelegateCreation);
+            validator.ValidateTag("pointer", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
+
+        [TestMethod]
+        public void DynamicObjectCreation_SetsNotNull()
+        {
+            const string code = @"
+var s = new Sample(dynamicArg);
+Tag(""s"", s);";
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:521]: That's why I added the `ValidateContainsOperation` to make sure it behaves as expected.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:521]: These two are `ObjectCreation`. And their results are than passed to a regular assignment => not relevant here.
```
dynamic d = new object(); // or even better:
dynamic d = new ExpandoObject();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 7]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+            validator.ValidateTag("s", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
+
+        [TestMethod]
+        public void ObjectCreation_SetsNotNull()
+        {
+            const string code = @"
+object assigned;
+var obj = new Object();
+var declared = new Exception();
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:533]: Just to make sure that the `System.Object`  doesn't behave in any funny way.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 8]
```diff
@@ -466,5 +466,97 @@ public void FlowCapture_SetsCapture()
...
+            validator.ValidateTag("object", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
+
+        [TestMethod]
+        public void TypeParameterObjectCreation_SetsNotNull()
+        {
+            const string code = @"
+public void Main<T>() where T : new()
+{
+    var value = new T();
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:552]: Why not? `Null` is definitely invalid and negation is `NotNull`. `value.Something` is safe to do so having `NotNull` is correct.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:556]: That's irrelevant. The `Tag` is only scaffolding noise that is not available when `CreateCSMethod` is called, contrary to `CreateCS` that provides it.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 3 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:260]: I was wondering whether there are other OperationKindEx values with "creation" in the name. I checked and these are all creations.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 9]
```diff
@@ -244,14 +244,20 @@ semantics switch
...
                 _ => context.State
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:261]: What about `OpartionKindEx.With`? This is also creating a new object behind the scenes:
```cs
Record R(int I);
var r1 = new R(1);
r2 = r1 with { I = 2 }; // calls new R(r1)
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:261]: Noted for another PR

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Comment by martin-strecker-sonarsource** [comment]
LGTM (merged).

**Pull request merged**